src/Adapter/Amp/EventLoop.php

```diff
@@ -79,27 +84,40 @@ public function async(\Generator $generator): Promise
         $deferred = new \Amp\Deferred();
         \Amp\Promise\rethrow(new \Amp\Coroutine($wrapper($generator, $deferred)));
 
-        return Internal\PromiseWrapper::createUnhandled($deferred->promise(), $this->unhandledFailingPromises);
+        $cancellable = function () use (&$currentPromise) {
+            $currentPromise->cancel();
+        };
+
+        return Internal\PromiseWrapper::createUnhandled($deferred->promise(), $this->unhandledFailingPromises, $cancellable);
     }
 
     /**
      * {@inheritdoc}
      */
     public function promiseAll(Promise ...$promises): Promise
     {
+        $cancellable = function () use (&$promises) {
+            foreach ($promises as $promise) {
+                $promise->cancel();
+            }
+        };
```

Comment:
This behavior should be clearly documented in function DocBlock: _If you cancel the returned promise, every sub-promise will be cancelled too_ 
Same for every function